### PR TITLE
[Enhancement] Add trace log for MV Rewrite to better debug

### PR DESCRIPTION
**Pull request by LiShuMing**
Introduce mv trace log to better debug mv rewrite.

Further:
- We may add more informations for MV Rewrite.

### Case1: MV is outdated
```
./fe.log:2023-06-16 14:49:41,155 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVPrepare():136] [MV TRACE] [PREPARE f81a89b5-0c11-11ee-98b1-162d24e49988] No Related Async MVs for plan
./fe.log:2023-06-16 14:49:41,155 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVPrepare():136] [MV TRACE] [PREPARE f81a89b5-0c11-11ee-98b1-162d24e49988] No Related Sync MVs
./fe.log:2023-06-16 14:49:41,155 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVPrepare():136] [MV TRACE] [PREPARE f81a89b5-0c11-11ee-98b1-162d24e49988] There are 0 candidate MVs after prepare phase
./fe.log:2023-06-16 14:49:42,716 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVPrepare():136] [MV TRACE] [PREPARE f908ba46-0c11-11ee-98b1-162d24e49988] MV user_tags_mv1 is outdated, partitionNamesToRefresh:[user_tags_mv1]
./fe.log:2023-06-16 14:49:42,716 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVPrepare():136] [MV TRACE] [PREPARE f908ba46-0c11-11ee-98b1-162d24e49988] RelatedMVs: [user_tags_mv1], CandidateMVs: []
./fe.log:2023-06-16 14:49:42,716 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVPrepare():136] [MV TRACE] [PREPARE f908ba46-0c11-11ee-98b1-162d24e49988] No Related Sync MVs
./fe.log:2023-06-16 14:49:42,716 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVPrepare():136] [MV TRACE] [PREPARE f908ba46-0c11-11ee-98b1-162d24e49988] There are 0 candidate MVs after prepare phase
```

### Case2: Expr cannot be rewritten

```
./fe.log:2023-06-16 14:49:10,603 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVPrepare():136] [MV TRACE] [PREPARE e5e4ac30-0c11-11ee-98b1-162d24e49988] Add MV user_tags_mv1 as a candidate
./fe.log:2023-06-16 14:49:10,603 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVPrepare():136] [MV TRACE] [PREPARE e5e4ac30-0c11-11ee-98b1-162d24e49988] RelatedMVs: [user_tags_mv1], CandidateMVs: [user_tags_mv1]
./fe.log:2023-06-16 14:49:10,603 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVPrepare():136] [MV TRACE] [PREPARE e5e4ac30-0c11-11ee-98b1-162d24e49988] No Related Sync MVs
./fe.log:2023-06-16 14:49:10,603 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVPrepare():136] [MV TRACE] [PREPARE e5e4ac30-0c11-11ee-98b1-162d24e49988] There are 1 candidate MVs after prepare phase
./fe.log:2023-06-16 14:49:10,604 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVRewrite():144] [MV TRACE] [REWRITE e5e4ac30-0c11-11ee-98b1-162d24e49988 TF_MV_AGGREGATE_SCAN_RULE user_tags_mv1] There are 1 relation id mappings to rewrite query
./fe.log:2023-06-16 14:49:10,605 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVRewrite():144] [MV TRACE] [REWRITE e5e4ac30-0c11-11ee-98b1-162d24e49988 TF_MV_AGGREGATE_SCAN_RULE user_tags_mv1] Rewrite aggregate group-by/agg expr failed: count(distinct add(cast(4: tag_id as bigint(20)), 1))
./fe.log:2023-06-16 14:49:10,605 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVRewrite():156] [MV TRACE] [REWRITE e5e4ac30-0c11-11ee-98b1-162d24e49988 TF_MV_AGGREGATE_SCAN_RULE] Generate 0 candidate plans after this rule rewrite
./fe.log:2023-06-16 14:49:10,605 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVRewrite():144] [MV TRACE] [REWRITE e5e4ac30-0c11-11ee-98b1-162d24e49988 TF_MV_ONLY_SCAN_RULE user_tags_mv1] MV is not applicable for this query: user_tags_mv1
./fe.log:2023-06-16 14:49:10,605 INFO (starrocks-mysql-nio-pool-2|268) [OptimizerTraceUtil.logMVRewrite():156] [MV TRACE] [REWRITE e5e4ac30-0c11-11ee-98b1-162d24e49988 TF_MV_ONLY_SCAN_RULE] Generate 0 candidate plans after this rule rewrite
```

### Case3: Rewrite Success
```
./fe.log:2023-06-16 14:47:55,933 INFO (starrocks-mysql-nio-pool-1|229) [OptimizerTraceUtil.logMVPrepare():136] [MV TRACE] [PREPARE b9616896-0c11-11ee-98b1-162d24e49988] Add MV user_tags_mv1 as a candidate
./fe.log:2023-06-16 14:47:55,933 INFO (starrocks-mysql-nio-pool-1|229) [OptimizerTraceUtil.logMVPrepare():136] [MV TRACE] [PREPARE b9616896-0c11-11ee-98b1-162d24e49988] RelatedMVs: [user_tags_mv1], CandidateMVs: [user_tags_mv1]
./fe.log:2023-06-16 14:47:55,934 INFO (starrocks-mysql-nio-pool-1|229) [OptimizerTraceUtil.logMVPrepare():136] [MV TRACE] [PREPARE b9616896-0c11-11ee-98b1-162d24e49988] No Related Sync MVs
./fe.log:2023-06-16 14:47:55,934 INFO (starrocks-mysql-nio-pool-1|229) [OptimizerTraceUtil.logMVPrepare():136] [MV TRACE] [PREPARE b9616896-0c11-11ee-98b1-162d24e49988] There are 1 candidate MVs after prepare phase
./fe.log:2023-06-16 14:47:55,949 INFO (starrocks-mysql-nio-pool-1|229) [OptimizerTraceUtil.logMVRewrite():144] [MV TRACE] [REWRITE b9616896-0c11-11ee-98b1-162d24e49988 TF_MV_AGGREGATE_SCAN_RULE user_tags_mv1] There are 1 relation id mappings to rewrite query
./fe.log:2023-06-16 14:47:55,955 INFO (starrocks-mysql-nio-pool-1|229) [OptimizerTraceUtil.logMVRewrite():144] [MV TRACE] [REWRITE b9616896-0c11-11ee-98b1-162d24e49988 TF_MV_AGGREGATE_SCAN_RULE user_tags_mv1] Rewrite Succeed:
./fe.log:2023-06-16 14:47:55,957 INFO (starrocks-mysql-nio-pool-1|229) [OptimizerTraceUtil.logMVRewrite():156] [MV TRACE] [REWRITE b9616896-0c11-11ee-98b1-162d24e49988 TF_MV_AGGREGATE_SCAN_RULE] Generate 1 candidate plans after this rule rewrite
./fe.log:2023-06-16 14:47:55,957 INFO (starrocks-mysql-nio-pool-1|229) [OptimizerTraceUtil.logMVRewrite():144] [MV TRACE] [REWRITE b9616896-0c11-11ee-98b1-162d24e49988 TF_MV_ONLY_SCAN_RULE user_tags_mv1] MV is not applicable for this query: user_tags_mv1
./fe.log:2023-06-16 14:47:55,957 INFO (starrocks-mysql-nio-pool-1|229) [OptimizerTraceUtil.logMVRewrite():156] [MV TRACE] [REWRITE b9616896-0c11-11ee-98b1-162d24e49988 TF_MV_ONLY_SCAN_RULE] Generate 0 candidate plans after this rule rewrite
./fe.log:2023-06-16 14:47:55,957 INFO (starrocks-mysql-nio-pool-1|229) [OptimizerTraceUtil.logMVRewrite():144] [MV TRACE] [REWRITE b9616896-0c11-11ee-98b1-162d24e49988 TF_MV_ONLY_SCAN_RULE user_tags_mv1] MV is not applicable for this query: user_tags_mv1
./fe.log:2023-06-16 14:47:55,957 INFO (starrocks-mysql-nio-pool-1|229) [OptimizerTraceUtil.logMVRewrite():156] [MV TRACE] [REWRITE b9616896-0c11-11ee-98b1-162d24e49988 TF_MV_ONLY_SCAN_RULE] Generate 0 candidate plans after this rule rewrite
```

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 78 / 104 (75.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|4|50.00%|[1909, 1910]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java|10|19|52.63%|[427, 447, 555, 590, 591, 611, 612, 618, 619]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|2|3|66.67%|[100]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|25|36|69.44%|[232, 237, 327, 363, 374, 1003, 1010, 1438, 1439, 1672, 1695]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|13|16|81.25%|[190, 191, 217]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceUtil.java|17|17|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewriteContext.java|3|3|100.00%|[]|

**Review by murphyatwork — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 1]
```diff
@@ -1129,7 +1159,7 @@ private ScalarOperator processNullPartition(RewriteContext rewriteContext, Scala
                 if (partitionRelatedPredicates.size() != 1) {
                     // has other partition predicates except partition column is null
                     // do not support now
-                    // can it happend?
+                    // can it happened?
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:1162]:
```suggestion
                    // can it happen?
```

**Review by ABingHuang — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**